authentik/stages/prompt/models.py

```diff
@@ -115,8 +136,46 @@ def serializer(self) -> Type[BaseSerializer]:
 
         return PromptSerializer
 
+    def get_choices(
+        self, prompt_context: dict, user: User, request: HttpRequest
+    ) -> Optional[tuple[dict[str, Any]]]:
+        """Get fully interpolated list of choices"""
+        if self.type not in CHOICE_FIELDS:
+            return None
+
+        raw_choices = self.placeholder
+
+        if self.field_key in prompt_context:
+            raw_choices = prompt_context[self.field_key]
+        elif self.placeholder_expression:
+            evaluator = PropertyMappingEvaluator(self, user, request, prompt_context=prompt_context)
+            try:
+                raw_choices = evaluator.evaluate(self.placeholder)
+            except Exception as exc:  # pylint:disable=broad-except
+                LOGGER.warning(
+                    "failed to evaluate prompt choices",
+                    exc=PropertyMappingExpressionException(str(exc)),
+                )
+
+        if isinstance(raw_choices, (list, tuple)):
+            choices = raw_choices
+        else:
+            choices = [raw_choices]
+
+        if len(choices) == 0:
+            LOGGER.warning("failed to get prompt choices", choices=choices, input=raw_choices)
+
+        return tuple(choices)
+
     def get_placeholder(self, prompt_context: dict, user: User, request: HttpRequest) -> str:
         """Get fully interpolated placeholder"""
+        if self.type in CHOICE_FIELDS:
+            # Make sure to return a valid choice as placeholder
+            choices = self.get_choices(prompt_context, user, request)
+            if choices:
+                return choices[0]
+            return ""
```

Comment:
```suggestion
            if not choices:
                return ""
            return choices[0]
```

imo reads a bit better like this, but fine with either one